app/models/user.rb

```diff
@@ -38,17 +64,116 @@ def load_effective_permissions
+    "#{USER_PROJECT_ID_PREFIX}_user_#{id}"
+  end
+
+  def viewable_project_ids(context)
+    return GrdaWarehouse::Hud::Project.project_ids_viewable_by(self, permission: context) if Rails.env.test?
+
+    Rails.cache.fetch("#{user_project_id_prefix}_#{context}", expires_in: EXPIRY_MINUTES.minutes) do
+      GrdaWarehouse::Hud::Project.project_ids_viewable_by(self, permission: context)
+    end
+  end
+
+  def editable_project_ids
+    return GrdaWarehouse::Hud::Project.project_ids_editable_by(self) if Rails.env.test?
+
+    Rails.cache.fetch("#{user_project_id_prefix}_editable", expires_in: EXPIRY_MINUTES.minutes) do
+      GrdaWarehouse::Hud::Project.project_ids_editable_by(self)
+    end
+  end
+
+  def self.clear_cached_permissions
+    Rails.cache.delete_matched("#{USER_PERMISSION_PREFIX}*")
+  end
+
+  def clear_cached_permissions
+    Rails.cache.delete_matched("#{user_permission_prefix}*")
+  end
+
+  def self.clear_cached_project_ids
+    Rails.cache.delete_matched("#{USER_PROJECT_ID_PREFIX}*")
+  end
+
+  def clear_cached_project_ids
+    Rails.cache.delete_matched("#{user_project_id_prefix}*")
+  end
+  # To fetch the list of Controlss that grant a user access to a particular set of projects
+  # user.access_group_for?('GrdaWarehouse::Hud::Project', 'can_view_projects')
+  # To fetch the list of Controlss that grant a user access to clients enrolled at as set of projects
+  # user.access_group_for?('GrdaWarehouse::Hud::Project', 'can_view_clients')
+  # def access_groups_for?(entity_type, perm)
+  #   return false unless entity_type.present? && perm.present?
+  #   return false unless send("#{perm}?")
+
+  #   role_ids = roles.where(perm => true).pluck(:id)
+  #   acs = access_controls.where(access_group_id: access_group_ids, role_id: role_ids)
+  #   FIXME, this isn't quite right yet
+  #   Controls.where(id: acs.pluck(:access_group_id), entity_type: entity_type)
+  # end
   def related_hmis_user(data_source)
     return unless HmisEnforcement.hmis_enabled?
 
     Hmis::User.find(id)&.tap { |u| u.update(hmis_data_source_id: data_source.id) }
   end
+
+  # list any cohort this user has some level of access to
+  def cohorts
+    # START_ACL cleanup after ACL migration is complete
+    if using_acls?
+      GrdaWarehouse::Cohort.where(id: collections.flat_map(&:cohort_ids))
+    else
+      GrdaWarehouse::Cohort.where(id: access_groups.flat_map(&:cohort_ids))
+    end
+    # END_ACL
+  end
+
+  # list any project groups the user has some level of access to
+  def project_groups
+    GrdaWarehouse::ProjectGroup.where(id: access_groups.flat_map(&:project_group_ids))
+  end
+
+  # list any data sources the user has some level of access to
+  def data_sources
+    GrdaWarehouse::DataSource.where(id: access_groups.flat_map(&:data_source_ids))
+  end
+
+  # list any reports the user has some level of access to
+  def reports
+    GrdaWarehouse::WarehouseReports::ReportDefinition.where(id: access_groups.flat_map(&:report_ids))
+  end
```

Comment:
Shouldn't these have an `if using_acls?` block that toggles looking at `collections` as opposed to `access_groups`, like the cohort function does?

Reply:
Maybe, but the concept will need to change because the relationship has to be within the context of a permission with ACLs.  That's definitely not fully implemented yet.

Reply:
I see, it looks like we're planning to remove usages of these functions after ACL migration is complete(?).
In the mean time though, something like `user.data_sources` like the usage below will be still using the users access_groups even if that user is set up to user acls. I guess thats OK, just wrapping my head around it. https://github.com/greenriver/hmis-warehouse/blob/eda76d9b3589c636b9df29ed30b48d18034e7da4/app/models/grda_warehouse/data_source.rb#L149-L159

Reply:
I went ahead and implemented these since I think it makes sense to enforce them even if slightly incorrectly for now.  Thanks!